### PR TITLE
Removed divides in Argon2

### LINKED ISSUE #5385: Get rid of modulo division in Argon2
See https://github.com/P-H-C/phc-winner-argon2/pull/306

We need this for both CPU and OpenCL formats. For CPU, perhaps do it after #2738.

### PR DESCRIPTION
**Pull request by Sc00bz**
Fixes #5385

### CODE REVIEW

**Comment by solardiz** [comment]
Thank you, @Sc00bz! I wonder if this makes a measurable performance difference - have you checked?

For the commit message, can you please prefix it with `Argon2: `? Amend and force-push.

Meanwhile, I've just let the CI jobs run on this PR.

File: src/argon2_opt_plug.c [hunk 1]
```diff
@@ -228,36 +230,47 @@ void argon2_fill_segment(const argon2_instance_t *instance,
...
-        if (curr_offset % instance->lane_length == 1) {
-            prev_offset = curr_offset - 1;
-        }
-
+         ++i, ++curr_offset) {
         /* 1.2 Computing the index of the reference block */
         /* 1.2.1 Taking pseudo-random value from the previous block */
         if (data_independent_addressing) {
-            if (i % ARGON2_ADDRESSES_IN_BLOCK == 0) {
+            if (i_address == 0) {
```
Comment (solardiz) [inline src/argon2_opt_plug.c]: I'm not sure we want this set of changes for `% ARGON2_ADDRESSES_IN_BLOCK`, which was likely just an AND. I suggest looking at the generated code with/without these changes.

**Comment by solardiz** [comment]
> Thank you, @Sc00bz! I wonder if this makes a measurable performance difference - have you checked?

We've just merged #5566, which will affect performance a lot. Ideally, the changes here would be benchmarked relative to and on top of that newer code.

**Comment by Sc00bz** [comment]
Fixed commit messages and undid the `i % ARGON2_ADDRESSES_IN_BLOCK` -> `i_address` change. I was not thinking.

**Review by solardiz — APPROVED** [review]
Thanks. I'm going to merge this, then experiment with it and maybe revise further.

**Comment by solardiz** [comment]
With this PR's changes as merged, there's slight increase in compiled code size (which is expected) and maybe a 1% or even 2% speedup at 1 thread on the old AMD FX-8120, same speed at 8 threads (where we probably bump into memory bandwidth anyway, although I didn't do the math to confirm). I'm actually a bit surprised there's measurable speedup - I don't know if it's a genuine effect of these changes or just some random fluctuation in how the old vs. new `john` binary is cached in memory (order of pages could matter, especially when cache associativity is low). But at least I saw no regression yet.

**Pull request merged**